lib/utils.js

```diff
@@ -1,18 +1,27 @@
 const WRITE = { CREATE: 1, UPDATE: 1, DELETE: 1 }
 
+const $hasPersonalData = Symbol('@cap-js/audit-logging:hasPersonalData')
+const $dataSubject = Symbol('@cap-js/audit-logging:dataSubject')
+const $parents = Symbol('@cap-js/audit-logging:parents')
+const $visitedUp = Symbol('@cap-js/audit-logging:visitedUp')
+const $visitedDown = Symbol('@cap-js/audit-logging:visitedDown')
+
 const hasPersonalData = entity => {
-  if (!entity['@PersonalData.EntitySemantics']) return
-  // default role to entity name
-  if (entity['@PersonalData.EntitySemantics'] === 'DataSubject' && !entity['@PersonalData.DataSubjectRole'])
-    entity['@PersonalData.DataSubjectRole'] = entity.name.match(/\w+/g).pop()
-  // prettier-ignore
-  const hasPersonalData = !!Object.values(entity.elements).some(element =>
-    element['@PersonalData.IsPotentiallyPersonal'] ||
-    element['@PersonalData.IsPotentiallySensitive'] ||
-    (element['@PersonalData.FieldSemantics'] && element['@PersonalData.FieldSemantics'] === 'DataSubjectID'))
-  // cache result
-  entity.own('_hasPersonalData', () => hasPersonalData)
-  return hasPersonalData
+  if (!entity.own($hasPersonalData)) {
+    if (!entity['@PersonalData.EntitySemantics']) entity.set($hasPersonalData, false)
+    else {
+      // default role to entity name
+      if (entity['@PersonalData.EntitySemantics'] === 'DataSubject' && !entity['@PersonalData.DataSubjectRole'])
+        entity['@PersonalData.DataSubjectRole'] = entity.name.match(/\w+/g).pop()
+      // prettier-ignore
+      const hasPersonalData = !!Object.values(entity.elements).some(element =>
```

Comment:
Not part of this change, but in general:
With universal CSN, annotations can also be part of the prototype chain

->

```js
for (const element in entity.elements) {/*...*/}
```